File: scripts/benchmark-huge.mjs
import { execFileSync } from "node:child_process";
import { createRequire } from "node:module";
import fs from "node:fs";
import path from "node:path";
import { pathToFileURL } from "node:url";
import { performance } from "node:perf_hooks";
import {
  clearEmbeddedBenchmarkCaches,
  embeddedOperationNames,
  runEmbeddedOperation,
} from "./embedded-language-benchmark.mjs";
import {
  benchmarkSourcesForRun,
  readBenchmarkCacheMode,
  readBenchmarkDisableCaches,
} from "./benchmark-cache-mode.mjs";

const require = createRequire(import.meta.url);
const root = path.resolve(import.meta.dirname, "..");
const sampleRoot = path.join(root, "samples", "classic-asp-huge-benchmark");
const generator = path.join(sampleRoot, "generate.mjs");
const coreDist = path.join(root, "packages", "core", "dist", "index.js");
const benchmarkIterations = readPositiveInteger("ASP_LSP_BENCH_ITERATIONS", 5);
const warmupIterations = readNonNegativeInteger("ASP_LSP_BENCH_WARMUPS", 1);
const benchmarkCacheMode = readBenchmarkCacheMode();
const disableCaches = readBenchmarkDisableCaches();
const operationFilter = process.env.ASP_LSP_BENCH_OPERATION;
const collectDebugSteps = readBoolean("ASP_LSP_BENCH_DEBUG_STEPS");
const analyzeStepTotals = new Map();
const results = [];

if (!fs.existsSync(coreDist)) {
  throw new Error(
    "packages/core/dist/index.js is missing. Run `pnpm --filter @asp-lsp/core run build`.",
  );
}

execFileSync(process.execPath, [generator], { stdio: "inherit" });

const core = require(coreDist);
const {
  analyzeVbscriptFromTextAsync,
  buildVirtualDocuments,
  collectVbscriptSymbolsFromTextAsync,
  parseAspDocumentAsync,
} = core;

const sources = collectBenchmarkSources();
const sourceStats = summarizeSources(sources);

if (shouldRunOperation("parseAspDocument")) {
  await runBenchmark("parseAspDocument", async (run) => {
    for (const source of sourcesForRun("parseAspDocument", run)) {
      await runSourceBenchmark(() => parseAspDocumentAsync(source.uri, source.text));
    }
  });
}

if (shouldRunOperation("buildVirtualDocuments")) {
  await runBenchmark("buildVirtualDocuments", async (run) => {
    for (const source of sourcesForRun("buildVirtualDocuments", run)) {
      await runSourceBenchmark(async () => {
        const parsed = await parseAspDocumentAsync(source.uri, source.text);
        buildVirtualDocuments(parsed);
      });
    }
  });
}

if (shouldRunOperation("collectVbscriptSymbols")) {
  await runBenchmark("collectVbscriptSymbols", async (run) => {
    for (const source of sourcesForRun("collectVbscriptSymbols", run)) {
      await runSourceBenchmark(() => collectVbscriptSymbolsFromTextAsync(source.uri, source.text));
    }
  });
}

if (shouldRunOperation("analyzeVbscript")) {
  await runBenchmark("analyzeVbscript", async (run) => {
    for (const source of sourcesForRun("analyzeVbscript", run)) {
      await runSourceBenchmark(() =>
        analyzeVbscriptFromTextAsync(source.uri, source.text, {}, analyzeContext()),
      );
    }
  });
}

for (const operation of embeddedOperationNames) {
  if (!shouldRunOperation(operation)) {
    continue;
  }
  await runBenchmark(operation, async (run) => {
    for (const source of sourcesForRun(operation, run)) {
      await runSourceBenchmark(() => runEmbeddedOperation(operation, source, core));
    }
  });
}

console.log("");
console.log(`Huge Classic ASP benchmark`);
console.log(`Files: ${sourceStats.files}`);
console.log(`Lines: ${sourceStats.lines.toLocaleString("en-US")}`);
console.log(`Bytes: ${sourceStats.bytes.toLocaleString("en-US")}`);
console.log(`Cache mode: ${benchmarkCacheMode}`);
console.log(`Benchmark caches: ${disableCaches ? "disabled" : "enabled"}`);
console.log(`Warmups: ${warmupIterations}`);
console.log(`Iterations: ${benchmarkIterations}`);
if (operationFilter) {
  console.log(`Operation filter: ${operationFilter}`);
}
console.log("");
printTable(results);
if (collectDebugSteps) {
  console.log("");
  console.log("analyzeVbscript debug step totals");
  console.log(
    `Measured calls include ${warmupIterations} warmup and ${benchmarkIterations} benchmark iterations.`,
  );
  console.log("");
  printDebugStepTotals(analyzeStepTotals);
}

function sourcesForRun(operation, run) {
  return benchmarkSourcesForRun(sources, benchmarkCacheMode, operation, run);
}

function shouldRunOperation(operation) {
  return !operationFilter || operationFilter === operation;
}

async function runSourceBenchmark(action) {
  if (!disableCaches) {
    return action();
  }
  clearBenchmarkCaches();
  try {
    return await action();
  } finally {
    clearBenchmarkCaches();
  }
}

function clearBenchmarkCaches() {
  core.clearAspCoreCaches?.();
  clearEmbeddedBenchmarkCaches();
}

function collectBenchmarkSources() {
  const relativePaths = [
    "default.asp",
    "includes/layer1.inc",
    "includes/layer2.inc",
    "includes/layer3.inc",
    "includes/layer4.inc",
    ...fs
      .readdirSync(path.join(sampleRoot, "includes", "generated"))
      .filter((entry) => entry.endsWith(".inc"))
      .sort()
      .map((entry) => path.join("includes", "generated", entry)),
  ];
  return relativePaths.map((relativePath) => {
    const absolutePath = path.join(sampleRoot, relativePath);
    return {
      relativePath,
      uri: pathToFileURL(absolutePath).href,
      text: fs.readFileSync(absolutePath, "utf8"),
    };
  });
}

function summarizeSources(items) {
  return items.reduce(
    (stats, item) => {
      stats.files += 1;
      stats.lines += item.text.split("\n").length - 1;
      stats.bytes += Buffer.byteLength(item.text);
      return stats;
    },
    { files: 0, lines: 0, bytes: 0 },
  );
}

async function runBenchmark(name, fn) {
  for (let index = 0; index < warmupIterations; index += 1) {
    await fn({ phase: "warmup", index });
  }

  const samples = [];
  for (let index = 0; index < benchmarkIterations; index += 1) {
    const start = performance.now();
    await fn({ phase: "measure", index });
    samples.push(performance.now() - start);
  }

  samples.sort((left, right) => left - right);
  const total = samples.reduce((sum, value) => sum + value, 0);
  results.push({
    name,
    min: samples[0],
    median: samples[Math.floor(samples.length / 2)],
    mean: total / samples.length,
    max: samples[samples.length - 1],
  });
}

function analyzeContext() {
  if (!collectDebugSteps) {
    return {};
  }
  return {
    debugStep(name, action) {
      const start = performance.now();
      const value = action();
      analyzeStepTotals.set(name, (analyzeStepTotals.get(name) ?? 0) + performance.now() - start);
      return value;
    },
  };
}

function printTable(items) {
  const rows = [
    ["Operation", "min ms", "median ms", "mean ms", "max ms"],
    ...items.map((item) => [
      item.name,
      formatMillis(item.min),
      formatMillis(item.median),
      formatMillis(item.mean),
      formatMillis(item.max),
    ]),
  ];
  const widths = rows[0].map((_, column) => Math.max(...rows.map((row) => row[column].length)));
  for (const [index, row] of rows.entries()) {
    console.log(row.map((value, column) => value.padEnd(widths[column])).join("  "));
    if (index === 0) {
      console.log(widths.map((width) => "-".repeat(width)).join("  "));
    }
  }
}

function printDebugStepTotals(totals) {
  const rows = [
    ["Step", "total ms"],
    ...[...totals.entries()]
      .sort((left, right) => right[1] - left[1])
      .map(([name, total]) => [name, formatMillis(total)]),
  ];
  const widths = rows[0].map((_, column) => Math.max(...rows.map((row) => row[column].length)));
  for (const [index, row] of rows.entries()) {
    console.log(row.map((value, column) => value.padEnd(widths[column])).join("  "));
    if (index === 0) {
      console.log(widths.map((width) => "-".repeat(width)).join("  "));
    }
  }
}

function formatMillis(value) {
  return value.toFixed(2);
}

function readPositiveInteger(name, fallback) {
  const raw = process.env[name];
  if (raw === undefined || raw === "") {
    return fallback;
  }
  const value = Number(raw);
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`${name} must be a positive integer.`);
  }
  return value;
}

function readNonNegativeInteger(name, fallback) {
  const raw = process.env[name];
  if (raw === undefined || raw === "") {
    return fallback;
  }
  const value = Number(raw);
  if (!Number.isInteger(value) || value < 0) {
    throw new Error(`${name} must be a non-negative integer.`);
  }
  return value;
}

function readBoolean(name) {
  const raw = process.env[name];
  if (raw === undefined || raw === "" || raw === "0" || raw.toLowerCase() === "false") {
    return false;
  }
  if (raw === "1" || raw.toLowerCase() === "true") {
    return true;
  }
  throw new Error(`${name} must be 1, 0, true, or false.`);
}

File: crates/asp-ide/examples/benchmark_huge.rs
use std::{
    fs,
    path::{Path, PathBuf},
    time::Instant,
};

use asp_ide::Ide;

struct Source {
    uri: String,
    text: String,
}

struct ResultRow {
    name: &'static str,
    samples: Vec<f64>,
}

fn main() -> Result<(), String> {
    let root = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .and_then(Path::parent)
        .ok_or_else(|| "workspace root not found".to_string())?
        .to_path_buf();
    let sample_root = root.join("samples").join("classic-asp-huge-benchmark");
    let mut sources = collect_sources(&sample_root)?;
    let iterations = read_usize("ASP_LSP_BENCH_ITERATIONS", 5)?;
    let warmups = read_usize("ASP_LSP_BENCH_WARMUPS", 1)?;
    if let Ok(raw) = std::env::var("ASP_LSP_BENCH_MAX_FILES") {
        if !raw.is_empty() {
            let max_files = raw
                .parse::<usize>()
                .map_err(|_| "ASP_LSP_BENCH_MAX_FILES must be a positive integer".to_string())?;
            if max_files == 0 {
                return Err("ASP_LSP_BENCH_MAX_FILES must be a positive integer".to_string());
            }
            sources.truncate(max_files);
        }
    }
    let operation_filter = std::env::var("ASP_LSP_BENCH_OPERATION").ok();
    let mut rows = Vec::new();

    if should_run(&operation_filter, "parseAspDocument") {
        rows.push(run_operation(
            "parseAspDocument",
            warmups,
            iterations,
            &sources,
            |ide, source| {
                let _ = ide.parse_asp(&source.uri)?;
                Ok(())
            },
        )?);
    }
    if should_run(&operation_filter, "buildVirtualDocuments") {
        rows.push(run_operation(
            "buildVirtualDocuments",
            warmups,
            iterations,
            &sources,
            |ide, source| {
                let _ = ide.embedded_virtual_documents(&source.uri)?;
                Ok(())
            },
        )?);
    }
    if should_run(&operation_filter, "collectVbscriptSymbols") {
        rows.push(run_operation(
            "collectVbscriptSymbols",
            warmups,
            iterations,
            &sources,
            |ide, source| {
                let _ = ide.vb_symbols(&source.uri)?;
                Ok(())
            },
        )?);
    }
    if should_run(&operation_filter, "analyzeVbscript") {
        rows.push(run_operation(
            "analyzeVbscript",
            warmups,
            iterations,
            &sources,
            |ide, source| {
                let _ = ide.vb_diagnostics(&source.uri)?;
                Ok(())
            },
        )?);
    }
    for name in [
        "htmlVirtualDocument",
        "cssVirtualDocument",
        "javascriptVirtualDocument",
    ] {
        if !should_run(&operation_filter, name) {
            continue;
        }
        rows.push(run_operation(
            name,
            warmups,
            iterations,
            &sources,
            |ide, source| {
                let language = name
                    .trim_end_matches("VirtualDocument")
                    .to_ascii_lowercase();
                let _ = ide
                    .embedded_virtual_documents(&source.uri)?
                    .into_iter()
                    .find(|document| document.document.language_id == language);
                Ok(())
            },
        )?);
    }

    println!();
    println!("Rust huge Classic ASP benchmark");
    println!("Files: {}", sources.len());
    println!("Warmups: {warmups}");
    println!("Iterations: {iterations}");
    println!();
    print_rows(&rows);
    Ok(())
}

fn should_run(filter: &Option<String>, name: &str) -> bool {
    filter.as_deref().is_none_or(|filter| filter == name)
}

fn run_operation<F>(
    name: &'static str,
    warmups: usize,
    iterations: usize,
    sources: &[Source],
    operation: F,
) -> Result<ResultRow, String>
where
    F: Fn(&Ide, &Source) -> Result<(), String>,
{
    for _ in 0..warmups {
        run_once(sources, &operation)?;
    }
    let mut samples = Vec::new();
    for _ in 0..iterations {
        samples.push(run_once(sources, &operation)?);
    }
    samples.sort_by(|left, right| left.total_cmp(right));
    Ok(ResultRow { name, samples })
}

fn run_once<F>(sources: &[Source], operation: &F) -> Result<f64, String>
where
    F: Fn(&Ide, &Source) -> Result<(), String>,
{
    let mut ide = Ide::default();
    for source in sources {
        ide.set_open_document(source.uri.clone(), source.text.clone());
    }
    let started = Instant::now();
    for source in sources {
        operation(&ide, source)?;
    }
    Ok(started.elapsed().as_secs_f64() * 1000.0)
}

fn collect_sources(sample_root: &Path) -> Result<Vec<Source>, String> {
    let mut relative_paths = vec![
        PathBuf::from("default.asp"),
        PathBuf::from("includes/layer1.inc"),
        PathBuf::from("includes/layer2.inc"),
        PathBuf::from("includes/layer3.inc"),
        PathBuf::from("includes/layer4.inc"),
    ];
    let generated = sample_root.join("includes").join("generated");
    let mut generated_entries = fs::read_dir(&generated)
        .map_err(|error| error.to_string())?
        .map(|entry| {
            entry
                .map_err(|error| error.to_string())
                .map(|entry| entry.path())
        })
        .collect::<Result<Vec<_>, _>>()?;
    generated_entries.sort();
    relative_paths.extend(generated_entries.into_iter().filter_map(|path| {
        (path.extension().and_then(|extension| extension.to_str()) == Some("inc")).then(|| {
            PathBuf::from("includes")
                .join("generated")
                .join(path.file_name().unwrap_or_default())
        })
    }));
    relative_paths
        .into_iter()
        .map(|relative| {
            let absolute = sample_root.join(&relative);
            let text = fs::read_to_string(&absolute).map_err(|error| error.to_string())?;
            Ok(Source {
                uri: format!("file://{}", absolute.to_string_lossy()),
                text,
            })
        })
        .collect()
}

fn read_usize(name: &str, fallback: usize) -> Result<usize, String> {
    match std::env::var(name) {
        Ok(raw) if !raw.is_empty() => raw
            .parse::<usize>()
            .map_err(|_| format!("{name} must be a positive integer"))
            .and_then(|value| {
                if value > 0 {
                    Ok(value)
                } else {
                    Err(format!("{name} must be a positive integer"))
                }
            }),
        _ => Ok(fallback),
    }
}

fn print_rows(rows: &[ResultRow]) {
    println!(
        "{:<28} {:>8} {:>10} {:>9} {:>8}",
        "Operation", "min ms", "median ms", "mean ms", "max ms"
    );
    println!("{:-<28} {:-<8} {:-<10} {:-<9} {:-<8}", "", "", "", "", "");
    for row in rows {
        let total = row.samples.iter().sum::<f64>();
        let mean = total / row.samples.len() as f64;
        println!(
            "{:<28} {:>8.2} {:>10.2} {:>9.2} {:>8.2}",
            row.name,
            row.samples[0],
            row.samples[row.samples.len() / 2],
            mean,
            row.samples[row.samples.len() - 1]
        );
    }
}
